alphadia/workflow/peptidecentric.py

```diff
@@ -444,89 +405,205 @@ def calibration(self):
+            )
+        if self.dia_data.has_ms1:
+            if self.config["search"]["target_ms1_tolerance"] > 0:
+                ms1_optimizer = searchoptimization.TargetedMS1Optimizer(
+                    self.config["search_initial"]["initial_ms1_tolerance"],
+                    self.config["search"]["target_ms1_tolerance"],
+                    self,
+                )
             else:
-                break
+                ms1_optimizer = searchoptimization.AutomaticMS1Optimizer(
+                    self.config["search_initial"]["initial_ms1_tolerance"],
+                    self,
+                )
+        else:
+            ms1_optimizer = None
+        if self.dia_data.has_mobility:
+            if self.config["search"]["target_mobility_tolerance"] > 0:
+                mobility_optimizer = searchoptimization.TargetedMobilityOptimizer(
+                    self.config["search_initial"]["initial_mobility_tolerance"],
+                    self.config["search"]["target_mobility_tolerance"],
+                    self,
+                )
+            else:
+                mobility_optimizer = searchoptimization.AutomaticMobilityOptimizer(
+                    self.config["search_initial"]["initial_mobility_tolerance"],
+                    self.calibration_manager,
+                    self.com,
+                    self.fdr_manager,
+                )
+        else:
+            mobility_optimizer = None
+
+        optimizers = [
+            ms2_optimizer,
+            rt_optimizer,
+            ms1_optimizer,
+            mobility_optimizer,
+        ]
+        targeted_optimizers = [
+            [
+                optimizer
+                for optimizer in optimizers
+                if isinstance(optimizer, searchoptimization.TargetedOptimizer)
+            ]
+        ]
+        automatic_optimizers = [
+            [optimizer]
+            for optimizer in optimizers
+            if isinstance(optimizer, searchoptimization.AutomaticOptimizer)
+        ]
+
+        order_of_optimization = (
+            targeted_optimizers + automatic_optimizers
+            if any(targeted_optimizers)
+            else automatic_optimizers
+        )
+
+        self.reporter.log_string(
+            "Starting initial classifier training and precursor identification.",
+            verbosity="progress",
+        )
+
+        self.eg_idxes_for_calibration, precursor_df, fragments_df = (
+            self.extract_optimization_data(
+                self.config["calibration"]["min_precursors_for_optimization"],
+                self.config["calibration"]["min_training_iterations"],
+            )
+        )
+
+        self.com.fit({"classifier_version": self.fdr_manager.current_version})
+
+        precursor_df_filtered, fragments_df_filtered = self.filter_dfs(
+            precursor_df, fragments_df
+        )
 
-            self.start_of_epoch(current_epoch)
+        self.recalibration(precursor_df_filtered, fragments_df_filtered)
```

Comment:
Just to understand it correctly, this is the last calibration as part of getting the optimization lock?